src/main/java/org/ow2/proactive/connector/iaas/cloud/provider/jclouds/aws/AWSEC2JCloudsProvider.java

```diff
@@ -116,17 +118,10 @@
     }
 
     private InstanceCredentials createCredentialsIfNotExist(Infrastructure infrastructure, Instance instance) {
-        KeyPairApi keyPairApi = getKeyPairApi(infrastructure);
-        String infrastructureKeyPairName = infrastructure.getId();
-        // check whether a key pair for this infrastructure exists already
-        Set<KeyPair> keyPairs = keyPairApi.describeKeyPairsInRegion(getRegionFromImage(instance));
-        boolean infrastructureKeyPairExistsAlready = keyPairs.stream()
-                                                             .anyMatch(keyPair -> keyPair.getKeyName()
-                                                                                         .equals(infrastructureKeyPairName));
-        if (!infrastructureKeyPairExistsAlready) {
-            generatedPrivateKey = createKeyPair(infrastructure, instance);
+        if (generatedKeyPair == null) {
+            generatedKeyPair = createKeyPair(infrastructure, instance);
```

Comment:
OK I didn't imagine the keyPair beeing a class variable. 
The problem is that it will create a unique keypair shared with all the registered AWS infrastructures.
In AWS the keypair is stored per region => we retrieve the region from the image => having multiple AWS nodesources with images from different regions will fail.
I think this `createCredentialsIfNotExist` method should also use `KeyPairApi` to check if the keypair actually exists on the targetted AWS region.

Reply:
Yes, I didn't see this problem coming. I will provide a fix by creating a key pair per region, and by checking if they are registered in AWS before using them.